src/rust/engine/src/intrinsics.rs

```diff
@@ -390,58 +390,68 @@ fn create_digest_to_digest(
   context: Context,
   args: Vec<Value>,
 ) -> BoxFuture<'static, NodeResult<Value>> {
-  let file_items = externs::collect_iterable(&args[0]).unwrap();
-  let digests: Vec<_> = file_items
-    .into_iter()
-    .map(|file_item| {
-      let path: String = externs::getattr(&file_item, "path").unwrap();
-      let store = context.core.store();
-      async move {
-        let path = RelativePath::new(PathBuf::from(path))
-          .map_err(|e| format!("The `path` must be relative: {:?}", e))?;
-
-        let (is_file_content, is_file_entry) = {
-          let gil = Python::acquire_gil();
-          let py = gil.python();
-          (
-            file_item.hasattr(py, "content").unwrap(),
-            file_item.hasattr(py, "file_digest").unwrap(),
-          )
-        };
-
-        if is_file_content {
-          let bytes =
-            bytes::Bytes::from(externs::getattr::<Vec<u8>>(&file_item, "content").unwrap());
-          let is_executable: bool = externs::getattr(&file_item, "is_executable").unwrap();
-
-          let digest = store.store_file_bytes(bytes, true).await?;
-          let snapshot = store
-            .snapshot_of_one_file(path, digest, is_executable)
-            .await?;
-          let res: Result<_, String> = Ok(snapshot.digest);
-          res
-        } else if is_file_entry {
-          let digest_obj = externs::getattr(&file_item, "file_digest")?;
-          let digest = Snapshot::lift_file_digest(&digest_obj)?;
-          let is_executable: bool = externs::getattr(&file_item, "is_executable").unwrap();
-          let snapshot = store
-            .snapshot_of_one_file(path, digest, is_executable)
-            .await?;
-          let res: Result<_, String> = Ok(snapshot.digest);
-          res
-        } else {
-          store
-            .create_empty_dir(path)
-            .await
-            .map_err(|e| format!("{:?}", e))
-        }
-      }
-    })
-    .collect();
+  let mut file_contents: Vec<(RelativePath, bytes::Bytes, bool)> = Vec::new();
+  let mut file_entries: Vec<(RelativePath, Digest, bool)> = Vec::new();
+  let mut empty_dirs: Vec<RelativePath> = Vec::new();
+  for file_item in externs::collect_iterable(&args[0]).unwrap() {
+    let gil = Python::acquire_gil();
+    let py = gil.python();
```

Comment:
Consider holding the GIL for the whole loop by adding a block wrapping it?
```rust
{
  let gil = Python::acquire_gil();
  let py = gil.python();
  for file_item in ... {
    ...
  }
}

Reply:
True, which will make the PyO3 diff smaller